### PR TITLE
[clang-format]: Split alignment of declarations around assignment

### LINKED ISSUE #68079: [clang-format] Formatting regression for member function pointers using aligned assignment  and declaration
I've recently upgraded my tools from LLVM-16 to LLVM-17 and notices the following regression with clang-format.

.clang-format
```sh
AlignConsecutiveAssignments:  true
AlignConsecutiveDeclarations: true
```

test.cpp
```cpp
struct A {};

using Fn   = int (A::*)();
using RFn  = int (A::*)() &;
using RRFn = int (A::*)() &&;
```

results in the following formatting:
```cpp
struct A {};

using Fn   = int   (A::*)();
using RFn  = int  (A::*)() &;
using RRFn = int (A::*)() &&;

```

### PR DESCRIPTION
**Pull request by gedare**
Function pointers are detected as a type of declaration using FunctionTypeLParen. They are aligned based on rules for AlignConsecutiveDeclarations. When a function pointer is on the right-hand side of an assignment, the alignment of the function pointer can result in excessive whitespace padding due to the ordering of alignment, as the alignment processes a line from left-to-right and first aligns the declarations before and after the assignment operator, and then aligns the assignment operator. Injection of whitespace by alignment of declarations after the equal sign followed by alignment of the equal sign results in the excessive whitespace.

Fixes #68079.

### CODE REVIEW

File: clang/unittests/Format/FormatTest.cpp [hunk 1]
```diff
@@ -18783,6 +18783,15 @@ TEST_F(FormatTest, AlignConsecutiveDeclarations) {
                "                 \"bb\"};\n"
                "int   bbbbbbb = 0;",
                Alignment);
+  // http://llvm.org/PR68079
+  verifyFormat("using Fn   = int (A::*)();\n"
+               "using RFn  = int (A::*)() &;\n"
+               "using RRFn = int (A::*)() &&;",
+               Alignment);
+  verifyFormat("using Fn   = int    (A::*)();\n"
```
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/FormatTest.cpp:18791]: See https://github.com/llvm/llvm-project/issues/68079#issuecomment-1767116044

File: clang/lib/Format/WhitespaceManager.cpp [hunk 2]
```diff
@@ -973,13 +974,51 @@ void WhitespaceManager::alignConsecutiveShortCaseStatements() {
...
 
-void WhitespaceManager::alignConsecutiveDeclarations() {
+void WhitespaceManager::alignConsecutiveDeclarationsPreAssignment() {
   if (!Style.AlignConsecutiveDeclarations.Enabled)
     return;
 
   AlignTokens(
       Style,
       [](Change const &C) {
+        for (FormatToken *Prev = C.Tok->Previous; Prev; Prev = Prev->Previous)
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/WhitespaceManager.cpp:984]: Could you extract that check and use the same function for the alignment? I'd not like the code duplication.

**Review by HazardyKnusperkeks — APPROVED** [review]

**Comment by tru** [comment]
@gedare can you fix the merge conflict on this one?

**Comment by owenca** [comment]
> @gedare can you fix the merge conflict on this one?

We don't want to merge this yet. See [here](https://github.com/llvm/llvm-project/issues/68079#issuecomment-1776826021).

**Review by owenca — CHANGES_REQUESTED** [review]
Please see [here](https://github.com/llvm/llvm-project/issues/68079#issuecomment-1776826021).

**Comment by gedare** [comment]
force push was required to layer the revert commit correctly.

**Review by owenca — COMMENTED** [review]
See https://github.com/llvm/llvm-project/issues/66857#issuecomment-1793022188. Can you make this a suboption, e.g. `AlignFunctionPointers`? We can add another (e.g. `AlignConstructorsDestructors`) for ctors and dtors in the future.

**Comment by gedare** [comment]
> See [#66857 (comment)](https://github.com/llvm/llvm-project/issues/66857#issuecomment-1793022188). Can you make this a suboption, e.g. `AlignFunctionPointers`? We can add another (e.g. `AlignConstructorsDestructors`) for ctors and dtors in the future.

Yes, I will work on this.

**Review by mydeveloperday — APPROVED** [review]

**Comment by gedare** [comment]
@owenca this waits for re-review

**Comment by owenca** [comment]
> @owenca this waits for re-review

Can you rebase the patch?

**Comment by gedare** [comment]
> > @owenca this waits for re-review
> 
> Can you rebase the patch?

done

File: clang/include/clang/Format/Format.h [hunk 3]
```diff
@@ -225,6 +225,22 @@ struct FormatStyle {
     ///   bbb = 2;
     /// \endcode
     bool AlignCompound;
+    /// Only for ``AlignConsecutiveDeclarations``. Whether function pointers
+    /// are aligned.
```
Comment (owenca) [inline clang/include/clang/Format/Format.h:229]:
```suggestion
    /// Only for ``AlignConsecutiveDeclarations``. Whether function pointers are
    /// aligned.
```
Please regenerate the rst file after reflowing the comment.

File: clang/lib/Format/Format.cpp [hunk 4]
```diff
@@ -76,40 +76,49 @@ template <> struct MappingTraits<FormatStyle::AlignConsecutiveStyle> {
                 FormatStyle::AlignConsecutiveStyle(
                     {/*Enabled=*/false, /*AcrossEmptyLines=*/false,
                      /*AcrossComments=*/false, /*AlignCompound=*/false,
+                     /*AlignFunctionPointers=*/false,
                      /*PadOperators=*/true}));
```
Comment (owenca) [inline clang/lib/Format/Format.cpp:80]:
```suggestion
                     /*AlignFunctionPointers=*/false, /*PadOperators=*/true}));
```

File: clang/lib/Format/Format.cpp [hunk 5]
```diff
@@ -76,40 +76,49 @@ template <> struct MappingTraits<FormatStyle::AlignConsecutiveStyle> {
...
     IO.enumCase(Value, "Consecutive",
                 FormatStyle::AlignConsecutiveStyle(
                     {/*Enabled=*/true, /*AcrossEmptyLines=*/false,
                      /*AcrossComments=*/false, /*AlignCompound=*/false,
+                     /*AlignFunctionPointers=*/false,
                      /*PadOperators=*/true}));
```
Comment (owenca) [inline clang/lib/Format/Format.cpp:86]: Ditto.

File: clang/lib/Format/Format.cpp [hunk 6]
```diff
@@ -76,40 +76,49 @@ template <> struct MappingTraits<FormatStyle::AlignConsecutiveStyle> {
...
     IO.enumCase(Value, "AcrossEmptyLines",
                 FormatStyle::AlignConsecutiveStyle(
                     {/*Enabled=*/true, /*AcrossEmptyLines=*/true,
                      /*AcrossComments=*/false, /*AlignCompound=*/false,
+                     /*AlignFunctionPointers=*/false,
                      /*PadOperators=*/true}));
```
Comment (owenca) [inline clang/lib/Format/Format.cpp:92]: Ditto.

File: clang/lib/Format/Format.cpp [hunk 7]
```diff
@@ -76,40 +76,49 @@ template <> struct MappingTraits<FormatStyle::AlignConsecutiveStyle> {
...
+                                            /*AlignFunctionPointers=*/false,
+                                            /*PadOperators=*/true}));
 
     // For backward compatibility.
     IO.enumCase(Value, "true",
                 FormatStyle::AlignConsecutiveStyle(
                     {/*Enabled=*/true, /*AcrossEmptyLines=*/false,
                      /*AcrossComments=*/false, /*AlignCompound=*/false,
+                     /*AlignFunctionPointers=*/false,
                      /*PadOperators=*/true}));
```
Comment (owenca) [inline clang/lib/Format/Format.cpp:116]: Ditto.

File: clang/lib/Format/Format.cpp [hunk 8]
```diff
@@ -76,40 +76,49 @@ template <> struct MappingTraits<FormatStyle::AlignConsecutiveStyle> {
...
     IO.enumCase(Value, "false",
                 FormatStyle::AlignConsecutiveStyle(
                     {/*Enabled=*/false, /*AcrossEmptyLines=*/false,
                      /*AcrossComments=*/false, /*AlignCompound=*/false,
+                     /*AlignFunctionPointers=*/false,
                      /*PadOperators=*/true}));
```
Comment (owenca) [inline clang/lib/Format/Format.cpp:122]: Ditto.

File: clang/lib/Format/WhitespaceManager.cpp [hunk 9]
```diff
@@ -978,7 +978,14 @@ void WhitespaceManager::alignConsecutiveDeclarations() {
 
   AlignTokens(
       Style,
-      [](Change const &C) {
+      [&](Change const &C) {
+        if (Style.AlignConsecutiveDeclarations.AlignFunctionPointers) {
+          for (FormatToken *Prev = C.Tok->Previous; Prev; Prev = Prev->Previous)
```
Comment (owenca) [inline clang/lib/Format/WhitespaceManager.cpp:983]:
```suggestion
          for (const auto *Prev = C.Tok->Previous; Prev; Prev = Prev->Previous)
```

**Review by owenca — COMMENTED** [review]
Please also update the release notes. LGTM otherwise.

**Comment by gedare** [comment]
> Please also update the release notes. LGTM otherwise.

Fixed the comments and added a release note.

**Review by owenca — APPROVED** [review]

File: clang/docs/ReleaseNotes.rst [hunk 10]
```diff
@@ -1079,6 +1079,7 @@ clang-format
 - Add ``ObjCPropertyAttributeOrder`` which can be used to sort ObjC property
   attributes (like ``nonatomic, strong, nullable``).
 - Add ``.clang-format-ignore`` files.
+- Add ``AlignFunctionPointers`` sub-option for ``AlignConsecutiveDeclarations''.
```
Comment (owenca) [inline clang/docs/ReleaseNotes.rst:1082]:
```suggestion
- Add ``AlignFunctionPointers`` sub-option for ``AlignConsecutiveDeclarations``.
```

**Pull request merged**